neuralpp/symbolic/basic_expression.py

```diff
@@ -1,88 +1,58 @@
 from __future__ import annotations
-
 import operator
-import builtins
-from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, AtomicExpression, \
-    VariableNotTypedError, ExpressionType, get_arithmetic_function_type_from_argument_types, Context, \
-    QuantifierExpression, AbelianOperation
 from abc import ABC
-from typing import Any, List, Optional, Callable, Dict, get_args
+from typing import List, Optional, Callable, Dict, get_args
+
+from neuralpp.symbolic.error import VariableNotTypedError
+from neuralpp.symbolic.expression import (
+    AbelianOperation,
+    AtomicExpression,
+    Constant,
+    Context,
+    Expression,
+    FunctionApplication,
+    QuantifierExpression,
+    Variable,
+)
 import neuralpp.symbolic.functions as functions
-
-
-class AmbiguousTypeError(TypeError, ValueError):
-    def __init__(self, python_callable: Callable):
-        super().__init__(f"{python_callable} is ambiguous.")
-
-
-def boolean_function_of_arity(arity: int) -> Callable:
```

Comment:
I moved these extra functions that do not depend on `Expression` to the util folder. I had to remove a function from that folder due to circular dependency.